common/domain/handler.go

```diff
@@ -427,27 +427,284 @@ func (d *handlerImpl) UpdateDomain(
+	notificationVersion int64,
+) (*types.UpdateDomainResponse, error) {
+
+	// intendedDomainState will be modified
+	// into the intended shape by the functions here
+	intendedDomainState := currentState.DeepCopy()
+
+	configVersion := currentState.ConfigVersion
+	failoverVersion := currentState.FailoverVersion
+	failoverNotificationVersion := currentState.FailoverNotificationVersion
+	isGlobalDomain := currentState.IsGlobalDomain
+	gracefulFailoverEndTime := currentState.FailoverEndTime
+	currentActiveCluster := currentState.ReplicationConfig.ActiveClusterName
+	previousFailoverVersion := currentState.PreviousFailoverVersion
+	lastUpdatedTime := time.Unix(0, currentState.LastUpdatedTime)
+	wasActiveActive := currentState.ReplicationConfig.IsActiveActive()
+	now := d.timeSource.Now()
+
+	var activeClusterChanged bool
+	var configurationChanged bool
+
+	// Update replication config
+	replicationConfig, replicationConfigChanged, activeClusterChanged, err := d.updateReplicationConfig(
+		currentState.Info.Name,
+		intendedDomainState.ReplicationConfig,
+		updateRequest,
+	)
+	if err != nil {
+		return nil, err
+	}
+	if !activeClusterChanged && !replicationConfigChanged {
+		return nil, errInvalidFailoverNoChangeDetected
+	}
+
+	err = d.ensureUpdateOrFailoverCooldown(currentState)
+	if err != nil {
+		return nil, err
+	}
+
+	// Handle graceful failover request
+	if updateRequest.FailoverTimeoutInSeconds != nil {
+		gracefulFailoverEndTime, previousFailoverVersion, err = d.handleGracefulFailover(
+			updateRequest,
+			replicationConfig,
+			currentActiveCluster,
+			gracefulFailoverEndTime,
+			failoverVersion,
+			activeClusterChanged,
+			isGlobalDomain,
+		)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// replication config is a subset of config,
+	configurationChanged = replicationConfigChanged
+
+	if err = d.domainAttrValidator.validateDomainConfig(intendedDomainState.Config); err != nil {
+		return nil, err
+	}
+
+	err = d.validateDomainReplicationConfigForUpdateDomain(replicationConfig, isGlobalDomain, configurationChanged, activeClusterChanged)
+	if err != nil {
+		return nil, err
+	}
+
+	// Check the failover cool down time
+	if lastUpdatedTime.Add(d.config.FailoverCoolDown(intendedDomainState.Info.Name)).After(now) {
+		d.logger.Debugf("Domain was last updated at %v, failoverCoolDown: %v, current time: %v.", lastUpdatedTime, d.config.FailoverCoolDown(intendedDomainState.Info.Name), now)
+		return nil, errDomainUpdateTooFrequent
+	}
+
+	// set the versions
+	if configurationChanged {
+		configVersion++
+	}
+
+	var failoverType constants.FailoverType = constants.FailoverTypeGrace
+
+	// Force failover cleans graceful failover state
+	if updateRequest.FailoverTimeoutInSeconds == nil {
+		failoverType = constants.FailoverTypeForce
+		gracefulFailoverEndTime = nil
+		previousFailoverVersion = constants.InitialPreviousFailoverVersion
+	}
+
+	// Cases:
+	// 1. active-passive domain's ActiveClusterName is changed
```

Comment:
I think we can unify the logic to handle these 3 cases, but I'm fine if you want to do it in a separate PR.

Reply:
yes, that's what I'm doing next